src/net/sourceforge/kolmafia/persistence/DebugDatabase.java

```diff
@@ -2583,6 +2577,92 @@ private static void checkPower(final int itemId, final boolean force) {
     }
   }
 
+  // Helper method to force normalize Concoction comparisons to [-1, 0, 1] before testing
+  private static int sgn(int value) {
+    return Integer.compare(value, 0);
+  }
+
+  // Helper method to append item id
+  private static String getIString(Concoction con) {
+    return "[" + con.getItemId() + "] " + con;
+  }
+
+  public static void checkConcoctions() {
+    // Code intended to verify that Concoction.compareTo() meets its contract.  Since the
+    // concoctions data is in a file and this is an expensive check, in terms of time,
+    // moved out of unit testing to here.
+    Concoction[] ids;
+    int maxIndex;
+    String msg;
+    int[][] result;
+    LockableListModel<Concoction> usables = ConcoctionDatabase.getUsables();
+    // size is all elements.  getSize is visible elements.
+    maxIndex = usables.size();
+    ids = new Concoction[maxIndex];
+    int i = 0;
+    for (Concoction con : usables) {
+      ids[i++] = con;
+    }
+    result = new int[maxIndex][maxIndex];
+    for (i = 0; i < maxIndex; ++i) {
+      for (int j = 0; j < maxIndex; ++j) {
+        result[i][j] = sgn(ids[i].compareTo(ids[j]));
+      }
+    }
+    // sgn(x.compareTo(y)) == -sgn(y.compareTo(x)
+    for (i = 0; i < maxIndex; ++i) {
+      for (int j = 0; j < maxIndex; ++j) {
+        msg =
+            "Failed comparing (quasi symmetry) "
+                + getIString(ids[i])
+                + " and "
+                + getIString(ids[j]);
+        if (result[i][j] != -result[j][i]) {
+          KoLmafia.updateDisplay(msg);
+        }
+      }
+    }
+    // tests the portion of the contract that says (x.compareTo(y)==0) == (x.equals(y))
+    for (i = 0; i < maxIndex; ++i) {
```

Comment:
A case could be made for combining the two i, j loops into one but I haven't done that yet.